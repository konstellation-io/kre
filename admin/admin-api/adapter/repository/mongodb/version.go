package mongodb

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"

	"github.com/konstellation-io/kre/admin/admin-api/adapter/config"
	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
	"github.com/konstellation-io/kre/admin/admin-api/domain/usecase"
	"github.com/konstellation-io/kre/admin/admin-api/domain/usecase/logging"
)

type VersionRepoMongoDB struct {
	cfg        *config.Config
	logger     logging.Logger
	collection *mongo.Collection
}

func NewVersionRepoMongoDB(
	cfg *config.Config,
	logger logging.Logger,
	client *mongo.Client,
) *VersionRepoMongoDB {
	collection := client.Database(cfg.MongoDB.DBName).Collection("versions")
	return &VersionRepoMongoDB{
		cfg,
		logger,
		collection,
	}
}

func (r *VersionRepoMongoDB) Create(userID string, newVersion *entity.Version) (*entity.Version, error) {
	newVersion.ID = primitive.NewObjectID().Hex()
	newVersion.CreationDate = time.Now().UTC()
	newVersion.CreationAuthor = userID
	newVersion.Status = entity.VersionStatusStopped
	res, err := r.collection.InsertOne(context.Background(), newVersion)
	if err != nil {
		return nil, err
	}

	newVersion.ID = res.InsertedID.(string)
	return newVersion, nil
}

func (r *VersionRepoMongoDB) GetByID(id string) (*entity.Version, error) {
	v := &entity.Version{}
	filter := bson.D{{"_id", id}}

	err := r.collection.FindOne(context.Background(), filter).Decode(v)
	if err == mongo.ErrNoDocuments {
		return v, usecase.ErrVersionNotFound
	}

	return v, err
}

func (r *VersionRepoMongoDB) GetByIDs(ids []string) ([]*entity.Version, []error) {
	ctx, _ := context.WithTimeout(context.Background(), 60*time.Second)
	var versions []*entity.Version
	cur, err := r.collection.Find(ctx, bson.M{"_id": bson.M{"$in": ids}})
	if err != nil {
		return versions, []error{err}
	}
	defer cur.Close(ctx)

	for cur.Next(ctx) {
		var v entity.Version
		err = cur.Decode(&v)
		if err != nil {
			return versions, []error{err}
		}
		versions = append(versions, &v)
	}

	return versions, nil
}

func (r *VersionRepoMongoDB) Update(version *entity.Version) error {
	_, err := r.collection.ReplaceOne(context.Background(), bson.M{"_id": version.ID}, version)
	if err != nil {
		return err
	}

	return nil
}

func (r *VersionRepoMongoDB) GetByRuntime(runtimeID string) ([]*entity.Version, error) {
	ctx, _ := context.WithTimeout(context.Background(), 60*time.Second)
	var versions []*entity.Version
	cur, err := r.collection.Find(ctx, bson.M{"runtimeId": runtimeID})
	if err != nil {
		return versions, err
	}
	defer cur.Close(ctx)

	for cur.Next(ctx) {
		var v entity.Version
		err = cur.Decode(&v)
		if err != nil {
			return versions, err
		}
		versions = append(versions, &v)
	}

	return versions, nil
}

func (r *VersionRepoMongoDB) SetHasDoc(ctx context.Context, versionID string, hasDoc bool) error {
	result, err := r.collection.UpdateOne(ctx, bson.M{"_id": versionID}, bson.M{"$set": bson.M{"hasDoc": hasDoc}})
	if err != nil {
		return err
	}

	if result.ModifiedCount != 1 {
		return usecase.ErrVersionNotFound
	}

	return nil
}

func (r *VersionRepoMongoDB) SetStatus(ctx context.Context, versionID string, status entity.VersionStatus) error {
	result, err := r.collection.UpdateOne(ctx, bson.M{"_id": versionID}, bson.M{"$set": bson.M{"status": status}})
	if err != nil {
		return err
	}

	if result.ModifiedCount != 1 {
		return usecase.ErrVersionNotFound
	}

	return nil
}
