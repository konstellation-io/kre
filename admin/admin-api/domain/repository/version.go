package repository

//go:generate mockgen -source=${GOFILE} -destination=$PWD/mocks/repo_${GOFILE} -package=mocks

import (
	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
)

type VersionRepo interface {
	Create(userID string, version *entity.Version) (*entity.Version, error)
	GetByID(id string) (*entity.Version, error)
	GetByIDs(ids []string) ([]*entity.Version, []error)
	Update(version *entity.Version) error
	GetByRuntime(runtimeID string) ([]*entity.Version, error)
}
