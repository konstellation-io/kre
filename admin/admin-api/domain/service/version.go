package service

//go:generate mockgen -source=${GOFILE} -destination=$PWD/mocks/service_${GOFILE} -package=mocks

import (
	"context"

	"github.com/konstellation-io/kre/admin/admin-api/domain/entity"
)

type VersionService interface {
	Start(context.Context, *entity.Runtime, *entity.Version) error
	Stop(context.Context, *entity.Runtime, *entity.Version) error
	Unpublish(*entity.Runtime, *entity.Version) error
	Publish(*entity.Runtime, *entity.Version) error
	UpdateConfig(*entity.Runtime, *entity.Version) error
}
