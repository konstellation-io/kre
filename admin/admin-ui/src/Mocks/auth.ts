import { AccessLevel } from 'Graphql/types/globalTypes';
import { loader } from 'graphql.macro';

const GetUserEmailQuery = loader('Graphql/queries/getMe.graphql');

export const usernameMock = {
  request: {
    query: GetUserEmailQuery
  },
  result: {
    data: {
      me: {
        email: 'user@konstellation.com',
        id: 'someId',
        accessLevel: AccessLevel.ADMIN
      }
    }
  }
};

export const unauthorizedUsernameMock = {
  request: {
    query: GetUserEmailQuery
  },
  result: {
    data: {
      me: {
        email: 'unauthorizedUser@konstellation.com',
        id: 'someId',
        accessLevel: AccessLevel.ADMIN
      }
    }
  },
  error: new Error('Unauthorized user')
};
