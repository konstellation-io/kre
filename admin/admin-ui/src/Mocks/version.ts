import { RuntimeStatus, VersionStatus } from 'Graphql/types/globalTypes';

import { loader } from 'graphql.macro';

const GetConfigurationVariablesQuery = loader(
  'Graphql/queries/getConfigurationVariables.graphql'
);
const GetVersionWorkflowsQuery = loader(
  'Graphql/queries/getVersionWorkflows.graphql'
);
const GetVersionNodeStatusQuery = loader(
  'Graphql/subscriptions/versionNodeStatus.graphql'
);
const NodeStatusSubscription = loader(
  'Graphql/subscriptions/versionNodeStatus.graphql'
);
const PublishVersionMutation = loader(
  'Graphql/mutations/publishVersion.graphql'
);
const StartVersionMutation = loader('Graphql/mutations/startVersion.graphql');
const StopVersionMutation = loader('Graphql/mutations/stopVersion.graphql');
const UnpublishVersionMutation = loader(
  'Graphql/mutations/unpublishVersion.graphql'
);

export const runtime = {
  id: 'runtimeId',
  name: 'runtimeName',
  description: 'Some Description',
  creationAuthor: { email: 'some@user.com' },
  status: RuntimeStatus.STARTED,
  __typename: 'Runtime'
};
export const version = {
  __typename: 'Version',
  id: 'versionId',
  name: 'versionName',
  description: 'versionDescription',
  status: VersionStatus.STARTED,
  creationDate: '2020/01/01',
  creationAuthor: {
    __typename: 'User',
    id: 'userId',
    email: 'user@domain.com'
  },
  publicationDate: '2020/01/01',
  publicationAuthor: {
    __typename: 'User',
    id: 'userId',
    email: 'user@domain.com'
  },
  config: {
    vars: [],
    completed: false
  }
};
export const confVarsMock = {
  request: {
    query: GetConfigurationVariablesQuery
  },
  result: {
    data: {
      version: {
        status: 'STARTED',
        config: {
          completed: true,
          vars: [
            {
              key: 'var2',
              value: 'value1',
              type: 'VARIABLE'
            },
            {
              key: 'var1',
              value: 'value2',
              type: 'VARIABLE'
            },
            {
              key: 'var3',
              value: 'value3',
              type: 'FILE'
            }
          ]
        }
      }
    }
  }
};

export const workflowsMock = {
  request: {
    query: GetVersionWorkflowsQuery
  },
  result: {
    data: {
      version: {
        name: 'version01',
        status: 'STARTED',
        configurationCompleted: true,
        workflows: [
          {
            name: 'workflow01',
            nodes: [
              {
                id: 'node01',
                name: 'one node',
                status: 'STARTED'
              },
              {
                id: 'node02',
                name: 'another node',
                status: 'STARTED'
              }
            ],
            edges: [
              {
                id: 'edge01',
                fromNode: 'node01',
                toNode: 'node02'
              }
            ]
          }
        ]
      }
    }
  }
};

export const errorMorkflowsMock = {
  request: {
    query: GetVersionWorkflowsQuery
  },
  error: 'Some error'
};

export const nodeStatus = {
  request: {
    query: GetVersionNodeStatusQuery
  },
  result: {
    data: {
      versionNodeStatus: [
        {
          date: '2020-01-01',
          nodeId: 'nodeId1',
          status: 'STARTED',
          message: 'some message 1'
        },
        {
          date: '2020-01-01',
          nodeId: 'nodeId2',
          status: 'STARTED',
          message: 'some message 2'
        }
      ]
    }
  }
};

export const nodeStatusMock = {
  request: {
    query: NodeStatusSubscription
  },
  result: {
    data: {
      versionNodeStatus: {
        date: '2020/01/01',
        nodeId: 'node02',
        status: 'STARTED',
        message: 'some message'
      }
    }
  }
};

export const publishVersionMock = {
  request: {
    query: PublishVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const startVersionMock = {
  request: {
    query: StartVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const stopVersionMock = {
  request: {
    query: StopVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};

export const unpublishVersionMock = {
  request: {
    query: UnpublishVersionMutation
  },
  result: {
    data: {
      version
    }
  }
};
