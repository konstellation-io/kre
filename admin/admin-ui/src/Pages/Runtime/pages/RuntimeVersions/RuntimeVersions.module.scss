@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  height: 100%;
  display: flex;
  flex-direction: row;
}

.content {
  height: 100%;
  display: flex;
  flex-direction: column;
}

.runtimeInfo {
  background-color: palette(lowlight, 900);
  padding: 3.5 * $grid-unit 3 * $grid-unit;

  .infoFields {
    display: flex;
    flex-direction: horizontal;
  }
}

.description {
  @include font-body;
  color: font-color(lowlight);

  margin-top: 3 * $grid-unit;
}

.activeVersion {
  padding: 8px 24px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  flex-wrap: wrap;
  background-color: palette(base, 300);
}

.versionTitle,
.versionName {
  @include font-body;
  color: font-color(light);
}

.versionTitle {
  line-height: 40px;
}

.versionCreation {
  @include font-caption;
  color: font-color(light);

  margin: 5px 25px;
}

.versionStatus {
  display: flex;
  flex-direction: row;
  align-items: center;
  margin: 25px auto 25px 29px;
}
.versionName {
  margin-right: 25px;
}

.greenCircle {
  width: 8px;
  height: 8px;
  margin-right: 10px;
  border-radius: 50%;
  background-color: $color-ok;
  flex-shrink: 0;
}

.secondRow {
  display: flex;
  flex-direction: row;
  flex-grow: 1;
  align-items: center;
  padding-right: 16px;
}

.versionList {
  @include listItems;

  display: relative;
  padding: 0 0 30px 0;
  overflow: overlay;
}
