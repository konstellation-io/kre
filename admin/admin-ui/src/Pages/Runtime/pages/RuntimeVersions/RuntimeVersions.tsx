import {
  GetVersionConfStatus_runtime,
  GetVersionConfStatus_versions
} from 'Graphql/queries/types/GetVersionConfStatus';

import IconEmail from '@material-ui/icons/Email';
import IconTime from '@material-ui/icons/AccessTime';
import InfoField from './components/InfoField/InfoField';
import PublishedVersionStatus from './components/PublishedVersionStatus/PublishedVersionStatus';
import React from 'react';
import Title from 'Components/Title/Title';
import VersionInfo from './components/VersionInfo/VersionInfo';
import { VersionStatus } from 'Graphql/types/globalTypes';
import { formatDate } from 'Utils/format';
import styles from './RuntimeVersions.module.scss';

type Props = {
  runtime: GetVersionConfStatus_runtime;
  versions: GetVersionConfStatus_versions[];
};

function RuntimeVersions({ runtime, versions }: Props) {
  const noVersions = versions.length === 0;
  const nPublishedVersions: number = versions.filter(
    version => version.status === VersionStatus.PUBLISHED
  ).length;

  const versionsComponents = versions.map(
    (version: GetVersionConfStatus_versions) => (
      <VersionInfo key={version.id} version={version} />
    )
  );

  return (
    <div className={styles.content}>
      <Title>Versions of runtime {runtime.name}</Title>
      <div className={styles.runtimeInfo}>
        <div className={styles.infoFields}>
          <InfoField
            field="CREATOR"
            Icon={IconEmail}
            value={runtime.creationAuthor.email}
          />
          <InfoField
            field="CREATED"
            Icon={IconTime}
            value={formatDate(new Date(runtime.creationDate), true)}
          />
        </div>
        <p className={styles.description}>{runtime.description}</p>
      </div>
      <PublishedVersionStatus
        noVersions={noVersions}
        nPublishedVersions={nPublishedVersions}
      />
      <div className={styles.versionList}>{versionsComponents}</div>
    </div>
  );
}

export default RuntimeVersions;
