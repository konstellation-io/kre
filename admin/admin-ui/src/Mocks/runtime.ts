import { runtime, version } from './version';

import { loader } from 'graphql.macro';

const GetRuntimesQuery = loader('Graphql/queries/getRuntimes.graphql');
const GetRuntimeAndVersionQuery = loader(
  'Graphql/queries/getRuntimeAndVersions.graphql'
);
const CreateRuntimeMutation = loader('Graphql/mutations/createRuntime.graphql');
const RuntimeCreatedSubscription = loader(
  'Graphql/subscriptions/runtimeCreated.graphql'
);

export const dashboardMock = {
  request: {
    query: GetRuntimesQuery
  },
  result: {
    data: {
      runtimes: [
        {
          id: '00001',
          name: 'Some Name',
          description: 'Some Description',
          status: 'STARTED',
          creationDate: '2019-11-28T15:28:01+00:00',
          creationAuthor: { email: 'some@user.com' },
          publishedVersion: {
            status: 'STARTED'
          }
        },
        {
          id: '00002',
          name: 'Some Other Name',
          description: 'Some Other Description',
          status: 'STARTED',
          creationDate: '2019-11-27T15:28:01+00:00',
          creationAuthor: { email: 'some@user.com' },
          publishedVersion: {
            status: 'STARTED'
          }
        },
        {
          id: '00003',
          name: 'Creating runtime',
          description: 'Some Description',
          status: 'CREATING',
          creationDate: '2019-11-27T15:28:01+00:00',
          creationAuthor: { email: 'some@user.com' },
          publishedVersion: {
            status: 'CREATING'
          }
        }
      ]
    }
  }
};

export const dashboardErrorMock = {
  request: {
    query: GetRuntimesQuery
  },
  error: new Error('cannot get runtimes')
};

export const addRuntimeMock = {
  request: {
    query: CreateRuntimeMutation,
    variables: { input: { name: 'New Runtime' } }
  },
  result: {
    data: {
      createRuntime: { name: 'some name' }
    }
  }
};

export const getRuntimeAndVersionsMock = {
  request: {
    query: GetRuntimeAndVersionQuery
  },
  result: {
    data: {
      runtime,
      version
    }
  }
};

export const getRuntimeAndVersionsErrorMock = {
  request: {
    query: GetRuntimeAndVersionQuery
  },
  error: new Error('cannot get runtime and versions')
};

export const runtimeCreatedMock = {
  request: {
    query: RuntimeCreatedSubscription
  },
  result: {
    data: {
      runtimeCreated: { id: 'some id', name: 'some name' }
    }
  }
};
