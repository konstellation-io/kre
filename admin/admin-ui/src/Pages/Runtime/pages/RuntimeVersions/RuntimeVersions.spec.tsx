import { runtime, version } from 'Mocks/version';

import React from 'react';
import RuntimeVersions from './RuntimeVersions';
import { shallow } from 'enzyme';

describe('RuntimeVersions', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <RuntimeVersions runtime={runtime} versions={[version, version]} />
    );
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
