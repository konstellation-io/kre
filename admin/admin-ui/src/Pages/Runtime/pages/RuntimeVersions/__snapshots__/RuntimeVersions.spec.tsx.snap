// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuntimeVersions matches snapshot 1`] = `
<div
  className="content"
>
  <Component>
    Versions of runtime 
    runtimeName
  </Component>
  <div
    className="runtimeInfo"
  >
    <div
      className="infoFields"
    >
      <InfoField
        Icon={
          Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "EmailIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        field="CREATOR"
        value="some@user.com"
      />
      <InfoField
        Icon={
          Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "AccessTimeIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        field="CREATED"
        value="Invalid date"
      />
    </div>
    <p
      className="description"
    >
      Some Description
    </p>
  </div>
  <PublishedVersionStatus
    nPublishedVersions={0}
    noVersions={false}
  />
  <div
    className="versionList"
  >
    <VersionInfo
      key="versionId"
      version={
        Object {
          "__typename": "Version",
          "config": Object {
            "completed": false,
            "vars": Array [],
          },
          "creationAuthor": Object {
            "__typename": "User",
            "email": "user@domain.com",
            "id": "userId",
          },
          "creationDate": "2020/01/01",
          "description": "versionDescription",
          "errors": Array [],
          "hasDoc": false,
          "id": "versionId",
          "name": "versionName",
          "publicationAuthor": Object {
            "__typename": "User",
            "email": "user@domain.com",
            "id": "userId",
          },
          "publicationDate": "2020/01/01",
          "status": "STARTED",
        }
      }
    />
    <VersionInfo
      key="versionId"
      version={
        Object {
          "__typename": "Version",
          "config": Object {
            "completed": false,
            "vars": Array [],
          },
          "creationAuthor": Object {
            "__typename": "User",
            "email": "user@domain.com",
            "id": "userId",
          },
          "creationDate": "2020/01/01",
          "description": "versionDescription",
          "errors": Array [],
          "hasDoc": false,
          "id": "versionId",
          "name": "versionName",
          "publicationAuthor": Object {
            "__typename": "User",
            "email": "user@domain.com",
            "id": "userId",
          },
          "publicationDate": "2020/01/01",
          "status": "STARTED",
        }
      }
    />
  </div>
</div>
`;
